src/test/java/com/aws/greengrass/clientdevices/auth/iot/IotAuthClientFake.java

```diff
@@ -98,14 +102,84 @@ public Optional<Certificate> getIotCertificate(String certificatePem) throws Inv
 
     @Override
     public boolean isThingAttachedToCertificate(Thing thing, Certificate certificate) {
+        if (Objects.isNull(certificate)) {
+            return false;
+        }
+        return isThingAttachedToCertificate(thing, certificate.getCertificateId());
+    }
+
+    @Override
+    public boolean isThingAttachedToCertificate(Thing thing, String certificateId) {
         if (thingToCerts.containsKey(thing.getThingName())) {
-            return thingToCerts.get(thing.getThingName()).contains(certificate.getCertificateId());
+            return thingToCerts.get(thing.getThingName()).contains(certificateId);
         }
         return false;
     }
 
+
+    public void attachThingToCore(Supplier<String> thingName) {
+        thingsAttachedToCore.add(thingName);
+    }
+
+    public void detachThingFromCore(Supplier<String> thingName) {
+        thingsAttachedToCore.remove(thingName);
+    }
+
     @Override
-    public Stream<Thing> getThingsAssociatedWithCoreDevice() {
-        return thingsAttachedToCore.stream().map(Thing::of);
+    public Stream<List<AssociatedClientDevice>> getThingsAssociatedWithCoreDevice() {
```

Comment:
This code below just simulates what the AWS SDK is doing for pagination. We could have used just a stream that creates lists but I wanted it to work as close as possible to the sdk given it lazily loads each page and we could simulate errors better